crates/worker/src/partition/state_machine/lifecycle/notify_invocation_response.rs

```diff
@@ -0,0 +1,187 @@
+    GetInvocationOutputCompletion, GetInvocationOutputResult, GetPromiseCompletion,
+    GetPromiseResult, SleepCompletion,
+};
+use tracing::error;
+
+pub struct OnNotifyInvocationResponse {
+    pub invocation_id: InvocationId,
+    pub invocation_epoch: InvocationEpoch,
+    pub status: InvocationStatus,
+    pub completion: Completion,
+}
+
+impl<'ctx, 's: 'ctx, S> CommandHandler<&'ctx mut StateMachineApplyContext<'s, S>>
+    for OnNotifyInvocationResponse
+where
+    S: journal_table_v1::JournalTable
+        + journal_table_v2::JournalTable
+        + InvocationStatusTable
+        + TimerTable
+        + FsmTable
+        + PromiseTable
+        + StateTable
+        + OutboxTable,
+{
+    async fn apply(self, ctx: &'ctx mut StateMachineApplyContext<'s, S>) -> Result<(), Error> {
+        let OnNotifyInvocationResponse {
+            invocation_id,
+            invocation_epoch,
+            status,
+            completion,
+        } = self;
+        let invocation_status = ctx.get_invocation_status(&invocation_id).await?;
+
+        // Verify that we need to ingest this
+        if !invocation_status.should_accept_completion(invocation_epoch, completion.entry_index) {
+            debug_if_leader!(
+                ctx.is_leader,
+                "Ignoring InvocationResponse epoch {} completion id {}",
+                invocation_epoch,
+                completion.entry_index
+            );
+            return Ok(());
+        }
+
+        // This code needs to be revisited once we remove Service Protocol <= V3, depending on what we still want to carry in InvocationResponse.
+
+        let command = ctx
+            .storage
+            .get_command_by_completion_id(invocation_id, completion.entry_index)
+            .await?;
+
+        if let Some(cmd) = command {
+            let entry: journal_v2::Entry = match cmd.command_type() {
+                CommandType::GetPromise => GetPromiseCompletion {
+                    completion_id: completion.entry_index,
+                    result: match completion.result {
+                        CompletionResult::Success(s) => GetPromiseResult::Success(s),
+                        CompletionResult::Failure(code, message) => {
+                            GetPromiseResult::Failure(journal_v2::Failure { code, message })
+                        }
+                        CompletionResult::Empty => {
+                            return Err(Error::BadCompletionVariantForInvocationResponse(
+                                CommandType::GetPromise,
+                                completion.entry_index,
+                                "Empty",
+                            ));
+                        }
+                    },
+                }
+                .into(),
+                CommandType::Sleep => SleepCompletion {
+                    completion_id: completion.entry_index,
+                }
```

Comment:
I am a little bit confused by the name of the struct `OnNotifyInvocationResponse` and that we are also completing sleeps here which should not be the result of an invocation response, should they?

Reply:
This is essentially defensive code.

In journal V1 we used to carry all notifications back through `InvocationResponse`, so here I handle all the corner case for journal v2. Once we get rid of the old protocol, good part of this code goes away and we'll be able to finally assume `InvocationResponse` is really used only for calls and nothing else.

Reply:
> v2 protocol sends SleepCompletions as a Completion?

For sleeps specifically, this cannot happen.

Let me write a mapping with all the possible messages in `OnNotifyInvocationResponse`